packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -426,6 +427,48 @@ export default createRule<Options, MessageId>({
       return false;
     }
 
+    // Checks whether a member expression is nullable or not regardless of it's previous node.
+    //  Example:
+    //  ```
+    //  // 'bar' is nullable if 'foo' is null.
+    //  // but this function checks regardless of 'foo' type, so returns 'true'.
+    //  declare const foo: { bar : { baz: string } } | null
+    //  foo?.bar;
+    //  ```
+    function isNullableOriginFromPrev(
+      node: TSESTree.MemberExpression | TSESTree.OptionalMemberExpression,
+    ): boolean {
+      const prevType = getNodeType(node.object);
+      const property = node.property;
+      if (prevType.isUnion() && isIdentifier(property)) {
+        const ownPropertyType = prevType.types
+          .map(type => checker.getTypeOfPropertyOfType(type, property.name))
+          .find(t => t);
```

Comment:
Mapping over the types in the union isn't quite right, and will cause issues in the following case. The checker API is smart enough to analyse unions to get the type of a property on that union.

```ts
interface Foo {
    a: number;
}

interface Bar {
    a: string | null;
}

declare const x: Foo | Bar;
```

```bash
# type = type of x;
$ type.types.map(t => checker.typeToString(checker.getTypeOfPropertyOfType(t, 'a')))

> ["number", "string | null"]
```
Eg with your current logic, this case will just check the `number` case, and will thus false-positive on the property.

If you instead just ask for the property on the union type itself:
```bash
# type = type of x;
$ checker.typeToString(checker.getTypeOfPropertyOfType(type, 'a'))

> "string | number | null"
```
It will work as expected. However if for some reason the base type is nullish, you need to also scrub the null from the union first:
```ts
interface Foo {
    a: number;
}

interface Bar {
    a: string | null;
}

declare const x: {t: Foo | Bar} | null;

x?.t.a;
```
```bash
# type = type of x?.t;
$ nonNullType = checker.getNonNullableType(type)
$ checker.typeToString(checker.getTypeOfPropertyOfType(nonNullType, 'a'))

> "string | number | null"
```

Reply:
Thanks for the review. I'll change it soon :)